osquery/process/windows/process_ops.cpp

```diff
@@ -88,12 +90,57 @@ uint32_t getGidFromSid(PSID sid) {
     auto toks = osquery::split(sidString, "-");
     gid = tryTo<uint32_t>(toks.at(toks.size() - 1), 10).takeOr(gid);
     LocalFree(sidString);
-
   } else if (ret == NERR_Success) {
-    gid = LPUSER_INFO_3(userBuff)->usri3_primary_group_id;
+    // Use NetUserGetLocalGroups to get a Local Group GID for this user
+    LPWSTR userName = LPUSER_INFO_0(userBuff)->usri0_name;
+    WORD level = 0;
+    DWORD flags = 0;
+    DWORD prefMaxLen = MAX_PREFERRED_LENGTH;
+    DWORD entriesRead = 0;
+    DWORD totalEntries = 0;
+    std::unique_ptr<BYTE[]> sidSmartPtr = nullptr;
+    PSID sidPtr = nullptr;
+
+    ret = NetUserGetLocalGroups(nullptr,
+                                userName,
+                                level,
+                                flags,
+                                (LPBYTE*)&userGroupsBuff,
+                                prefMaxLen,
+                                &entriesRead,
+                                &totalEntries);
+
+    if (ret == NERR_Success) {
+      LPLOCALGROUP_USERS_INFO_0 pTmpBuf;
+      DWORD i;
+
+      // A user often has more than one local group. We only return the first!
+      if (userGroupsBuff != NULL) {
+        // From group name to group SID to the RID (osquery concept of GID):
+        sidSmartPtr = getSidFromUsername(userGroupsBuff->lgrui0_name);
+        if (sidSmartPtr != nullptr) {
+          sidPtr = static_cast<PSID>(sidSmartPtr.get());
+          auto rid = getRidFromSid(sidPtr);
+          gid = rid;
+        }
+      }
+    }
+
+    // If none of the above worked, User may not have a Local Group
+    if (gid == static_cast<uint32_t>(-1)) {
+      // Fallback to using its RID from its USER_INFO_3 struct
+      NetApiBufferFree(userBuff); // free the level 0 buff from above
+      userInfoLevel = 3;
```

Comment:
I see this in the docs... Wonder if we should heed that advice?

> Note that the USER_INFO_4 structure supersedes the USER_INFO_3 structure on Windows XP and later. **It is recommended that applications use the USER_INFO_4 structure instead of the USER_INFO_3 structure with the NetUserAdd, NetUserSetInfo, and NetUserGetInfofunctions on Windows XP and later.**

https://docs.microsoft.com/en-us/windows/win32/api/lmaccess/ns-lmaccess-user_info_4

Reply:
This is code I moved down a few lines unchanged, as a fallback method. If I had to guess the original intent of using the `level 3` struct, it was probably because the lower level structs are smaller in memory if you're going to be doing 50,000 of them (as was the case for some Domain Controllers that couldn't run the `users` table in older versions of osquery).

But looking at level 4 and level 3, they're the exact same size, only now one of the DWORDs is a proper PSID type. I can make that change if it's what MSDN recommends.